[NavigationDrawer] Implemented opt-in ability to have a "sticky" view, …

jakerockland opened the pull request:
…presented behind top safe area, when a headerViewController is not provided.

This PR adds to the MDCBottomDrawerViewController the ability to specify that a "sticky" view be used when a `headerViewController` is not provided in order to prevent the bottom drawer content from conflicting with status bar when a header has not been provided, addressing #6434.

This prevents clients from having to implement an empty header, akin to what I had suggested in #9436, which is what we're having to do now as a workaround.

I've added basic flag propagation unit tests around these changes as unit tests. I have also verified manually that this fix works as a drop-in (actually this solution is better than our workaround, as it does not result in UI oddities when the empty header is presented in landscape mode, while retaining respect for corner radius in collapsed mode) replacement for our team's existing workaround, and have added a screencast on the internal issue for this ticket [b/123186711](http://b/123186711).

Hope this approach is a bit more along the lines of your feedback @yarneo 😄

jakerockland committed:
[NavigationDrawer] Implemented opt-in ability to have a sticky view presented behind top safe area, when a headerViewController is not provided.

jakerockland committed:
[NavigationDrawer] Implemented opt-in ability to have a sticky view presented behind top safe area, when a headerViewController is not provided.

jakerockland commented on components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:
```diff
@@ -1014,7 +1033,7 @@ - (BOOL)contentScrollsToReveal {
 - (CGFloat)topHeaderHeight {
   if (!self.headerViewController) {
-    return 0;
+    return MDCDeviceTopSafeAreaInset();
```
One note about this change in particular. I intentionally do not have this behavior modified based on the `shouldUseStickyStatusBar` flag, as it seems to me that this _should_ always be including the safe area inset in the `topHeaderHeight`, as the `contentHeaderHeight` accurately reflects that the height is `0` when there is no `headerViewController` whereas, if I understand correctly, this value should reflect that content height + the necessary safe area offset. I've also updated the accompanying (already existing) unit test to reflect this.

yarneo replied:
Do you believe this may cause a visual change for clients by updating this value? or was it not used?

jakerockland replied:
So I believe it will cause a _minor_ visual change, but not what I think would be considered a regression. 

This will only effect instances of the drawer where no header is being provided and, if I understand correctly, the only change this will cause in that instance is that it will cause the offset that determines when to animate removing the handle and corner rounding to take into account the top safe area, regardless of whether or not there is a header. In effect this will just mean that the animation happens slightly sooner. 

This seemed to me to be the behavior that should happen, but I'm happy to put this behind the `shouldUseStickyStatusBar` flag as well if that seems off to you, or for safety reasons (given that there may be some other visual change caused by this that I am unaware of).

jakerockland replied:
An updated thought that struck me about this aspect of this PR: Given that #9471 and #9476 also effect this component, it may be safest to merge this after those two are merged and I have the chance re-base + re-test that there were not any unexpected changes to the behavior here. 

This will also allow me to update this call-site to also use the added `topSafeAreaInset` property rather than `MDCDeviceTopSafeAreaInset`, and verify that the flagged sticky behavior is still as expected. 

If that is the case, do you think it would be possible to get this change in before the next release is cut? Happy to do whatever additional testing you think necessary to get that done as we have a few internal issues that would be helped along with these changes 😄🙏

yarneo replied:
sounds good :) Thanks for thinking this through

yarneo replied:
Im working to get the other changes in before the next release, this one may not make the cut off as the next release is being cut monday morning

yarneo commented on components/NavigationDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -204,6 +204,14 @@
+/**
+ A flag allowing clients to opt-in to generating a "sticky" view that is presented behind the status
+ bar, to prevent bottom drawer content from conflicting with the status bar content when an explicit
+ header has not been provided by the client (see Github issue #6434, or internal issue b/123186711,
```
there's no need to report an issue in the API comment if there isn't any outstanding bug after this change

yarneo replied:
"This flag allows clients to have the drawer content scroll below the status bar when no header is provided.

Defaults to NO.

note: This flag is only applicable when headerViewController is nil.
"

jakerockland replied:
SGTM, this description is much better—will update to use this comment block throughout 😄

yarneo commented on components/NavigationDrawer/src/MDCBottomDrawerPresentationController.h:
```diff
@@ -204,6 +204,14 @@
+ header has not been provided by the client (see Github issue #6434, or internal issue b/123186711,
+ for more context).
+ */
+@property(nonatomic, assign) BOOL shouldUseStickyStatusBar;
```
what happens in general when a client has a headerViewController and also sets the shouldUseStickyStatusBar flag? Is this unexpected behavior, or should it still work well?

jakerockland replied:
Good point. If the client has provided a `headerViewController`, setting `shouldUseStickyStatusBar` to YES is a no-op and the behavior should work as if the client had only provided the `headerViewController` and not modified the flag. 

I think it may be useful to make more explicit (given that some clients, like us, may have a default generator for the bottom drawer that takes in a nullable `headerViewController`) and I have modified your suggested comment block to make this even more explicit, but let me know if you think the addition is unnecessary. 😅

yarneo commented on components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:
```diff
@@ -828,6 +837,16 @@ - (UIScrollView *)scrollView {
 
+- (UIView *)topSafeAreaView {
+  if (!_topSafeAreaView) {
+    _topSafeAreaView = [UIView new];
```
nit: please instantiate using [[UIView alloc] init]

jakerockland replied:
Done! 👍

yarneo commented on components/NavigationDrawer/src/MDCBottomDrawerPresentationController.m:
```diff
@@ -367,8 +368,16 @@ - (void)bottomDrawerContainerViewControllerTopTransitionRatio:
+    // The transition ratio is adjusted if a "sticky" status bar view is being used, in place of a
+    // `headerViewController`, to prevent a presentation issue where the drawer's corner radius is 
+    // not kept in sync with the animation of the sticky view's expansion.
+    CGFloat adjustedTransitionRatio = transitionRatio;
```
Do you think maybe it's possible to avoid doing this patch? Potentially maybe if we update the transition ratio correctly when this behavior is opted in, in the containerViewController implementation? could it be related to us not going into this code:
```
 if (self.hasHeaderViewController) {
  BOOL contentReachesFullscreen = self.contentReachesFullscreen;	    if ([self.headerViewController

            respondsToSelector:@selector(updateDrawerHeaderTransitionRatio:)]) {
  if ([self.headerViewController	      if (self.shouldAlwaysExpandHeader) {
          respondsToSelector:@selector(updateDrawerHeaderTransitionRatio:)]) {	        [self.headerViewController updateDrawerHeaderTransitionRatio:headerTransitionToTop];
    if (self.shouldAlwaysExpandHeader) {	      } else {
      [self.headerViewController updateDrawerHeaderTransitionRatio:headerTransitionToTop];	        [self.headerViewController
    } else {	            updateDrawerHeaderTransitionRatio:contentReachesFullscreen ? headerTransitionToTop : 0];
      [self.headerViewController	      }
```

jakerockland replied:
Good point. This can be handled entirely within the `containerViewController`, that was an oversight on my part. Have updated accordingly!

jakerockland committed:
Revised documentation + moved transition adjustment into MDCBottomDrawerContainerViewController

jakerockland committed:
Revised documentation + moved transition adjustment into MDCBottomDrawerContainerViewController

jakerockland commented:
Thanks so much for all the feedback here @yarneo, I believe I've addressed everything in the additional commit I've added to this PR, but please let me know if there is anything I may have overlooked.

jakerockland committed:
[NavigationDrawer] Move off of MDCDeviceTopSafeAreaInset.

yarneo committed:
Update MDCNavigationDrawerScrollViewTests.m

yarneo committed:
Update MDCBottomDrawerContainerViewController.m

yarneo committed:
Update MDCBottomDrawerViewController.m

yarneo committed:
Update MDCBottomDrawerContainerViewController.m

yarneo committed:
Update MDCLayoutMetrics.h

yarneo committed:
Update MDCLayoutMetrics.m

yarneo committed:
Update MDCLayoutMetrics.h

yarneo committed:
Update MDCLayoutMetrics.m

yarneo committed:
Update MDCLayoutMetrics.m

jakerockland committed:
Update tests to use safeAreas if available.

jakerockland committed:
Use self.topSafeAreaInset instead of MDCDeviceTopSafeAreaInset

jakerockland committed:
Merge branch 'opt-in-sticky-status-bar-view' of github.com:jakerockland/material-components-ios into opt-in-sticky-status-bar-view

jakerockland commented:
@yarneo I re-tested this PR with all of the changes from #9476 as well as now-merged #9471 and did not notice any unexpected changes in behavior. 🎊 

Because of that, I've rebased the changes here to be based off of #9476, with the hope that if the test changes in #9476 make it ready for tomorrow's cut, this change may also be ready with it.
